File: Biblioteca PUI/samples/c++/TocarAnimalesPUI.cpp
/*
 * TocarAnimalesPui.cpp
 * Author: smarichal
 */
#include "Pui.h"
#include "UtilidadesSDL.h"
#include <AR/gsub.h>
#include <GL/gl.h>
#include <GL/glut.h>
			/*************************************************************************************************************************
			 * 									CONFIGURACIONES GENERALES DE LA APLICACION
			 *************************************************************************************************************************/
int mostrarVideo = 1;
int mostrarRadar = 1;

			/*************************************************************************************************************************
			 * 									CONFIGURACION DE RECONOCIMIENTO DE COLORES
			 *************************************************************************************************************************/

#define CANT_HISTORIA 4

int mostrarImagenDebug = 0;
int debugImg = 0;		//Parametro para hayObjetoColor. Si se le pasa 1 y se le pasa una ventana entonces va reproduciendo el ROI. Sino crea una ventana nueva pero detiene la aplicacion hasta tocar una tecla

int usarAlgoritmoRapido =1;

			/********************************************************************************************************************************
			 * 								CONFIGURACION DE ARTOOLKIT Y MARCADORES
			 ********************************************************************************************************************************/

#define MAX_MARCADORES 30

/* no se define el pipeline aqui, se usa la variable de ambiente ARTOOLKIT_CONFIG*/
char			*vconf = "";

int             thresh = 100;
int             count = 0;

char           *cparam_name    = "Data/camera_para.dat";
ARParam         cparam;

char                *config_name = "Data/tocarAnimalesPUI/letras.dat";
Pui* pui;

/**************************************************************************************************************************************
 * ************************************************************************************************************************************
 *  										VARIABLES DEL PROTOTIPO
 * ************************************************************************************************************************************
 * ************************************************************************************************************************************/

typedef struct {
	int idImagenSDL;
}ElementoEnPantalla;

/* Representacion de los elementosEnPantalla*/
ElementoEnPantalla elementosEnPantalla[MAX_MARCADORES];	//El indice el elemento coincide con el id del elemento que tambien coincide con el id del marcador artoolkit
int cantElementosCargados = 0;

int idObjetoPUIActual = -1; //id del objeto PUI que esta en pantalla, con este elemento deberia corresponderse la seleccion del usuario para obtener una respuesta exitosa
int idObjetoPUISeleccionado = -1; //id del objeto PUI que se ha seleccionado

int correspondenciaCorrecta = 0; //indica si el elemento del conjunto A y el del conjunto B
								 // que se seleccionaron se corresponden
int correspondenciaIncorrecta = 0;

int xNombreRelacion = 200;
int yNombreRelacion = 15;

int xImagen = 135;
int yImagen = 100;


int anchoPantalla = 655;
int altoPantalla = 700;

//Son los id de las imagenes actualmente desplegadas en la pantalla
int idSDLImagenPantalla = -1;


				/***************************************************************************************************************************************
				 * 												DECALRACIONES DE FUNCIONES AUXILIARES
				 ****************************************************************************************************************************************/
static void   init(void);
static void   cleanup(void);
static void   keyEvent( unsigned char key, int x, int y);
static void   mainLoop(void);
static void   draw();

int main(int argc, char **argv){
	pui = new Pui();
	printf("InitAR\n");

	init();
    printf("main loop\n");
    argMainLoop( NULL, keyEvent, mainLoop );
    printf("---------------------");
	return (0);
}


void proximoElemento(){
	if(idObjetoPUIActual == cantElementosCargados -1){
		idObjetoPUIActual = 0;
	}else{
		idObjetoPUIActual++;
	}
}

void actualizarPantalla(){
	if(idObjetoPUIActual!=-1){
		if(elementosEnPantalla[idObjetoPUIActual].idImagenSDL == idSDLImagenPantalla && idSDLImagenPantalla >= 0){
			printf("Se sigue con la misma imagen, no se hace nada");
		}else{
			if(idSDLImagenPantalla!=-1){
				OcultarImage(idSDLImagenPantalla,0);
			}
			if( (elementosEnPantalla[idObjetoPUIActual].idImagenSDL == -1)  ){ //Todavia no se cargo la imagen para este elemento
				printf("voy a desplegar la imagen: %s\n",pui->getImagenObjetoPUI(idObjetoPUIActual));
				elementosEnPantalla[idObjetoPUIActual].idImagenSDL = PlayImageConTitulo(pui->getImagenObjetoPUI(idObjetoPUIActual),pui->getNombreObjetoPUI(idObjetoPUIActual),xImagen,yImagen,0);
				idSDLImagenPantalla = elementosEnPantalla[idObjetoPUIActual].idImagenSDL; //Actualizo el id de la imagen mostrada actualmente
				printf("IdSDLA = %d\n",idSDLImagenPantalla);
			}else{
				mostrarImagen(elementosEnPantalla[idObjetoPUIActual].idImagenSDL,0);
				idSDLImagenPantalla = elementosEnPantalla[idObjetoPUIActual].idImagenSDL;
			}
		}

		if(correspondenciaCorrecta){
			OcultarZona(xNombreRelacion,yNombreRelacion,anchoPantalla - xNombreRelacion,yImagen - yNombreRelacion,0); //Borro la zona superior de la pantalla que es donde esta escrito el texto
			EscribirTexto(pui->getNombreRelacionObjetoPUI(idObjetoPUIActual),xNombreRelacion,yNombreRelacion,0);
			PlaySound(pui->getSonidoObjetoPUI(idObjetoPUIActual));
		}else if(correspondenciaIncorrecta){
			PlaySound(pui->getSonidoError());
		}

	}else{
		EscribirTexto("Prototipo - Estado Inicial",xNombreRelacion,yNombreRelacion,0);
	}
	//Actualizo la pantalla
	printf("Flip antes\n");
	flip();
	printf("Flip dsps\n");
}


static void   keyEvent( unsigned char key, int x, int y)
{
    /* quit if the ESC key is pressed */
    if( key == 0x1b ) {
        printf("*** %f (frame/sec)\n", (double)count/arUtilTimer());
        cleanUpSDL();
		pui->finish();
		delete pui;

        exit(0);
    }


    if( key == 'v' ) {
		mostrarVideo = 1 - mostrarVideo;
		if(!mostrarVideo){
			argCleanWindow(0,0);
		}
    }

    if( key == 'r' ) {
    	printf("Mostrar Radar\n");
		mostrarRadar = 1 - mostrarRadar;
		if(!mostrarRadar){
			argCleanWindow(2,1);
		}
	}

    if( key == SDLK_RETURN ) {
		printf("Proximo\n");
		proximoElemento();
		actualizarPantalla();

	}

    if( key == 't' ) {
        printf("count = %d\n",count);
        printf("timer = %f\n",arUtilTimer());
        printf("*** %f (frame/sec)\n", (double)count/arUtilTimer());
        printf("Enter new threshold value (current = %d): ", thresh);
        scanf("%d",&thresh); while( getchar()!='\n' );
        pui->setARThreshold(thresh);
        printf("\n");
        count = 0;
    }

    /* turn on and off the debug mode with right mouse */
    if( key == 'd' ) {
        printf("*** %f (frame/sec)\n", (double)count/arUtilTimer());
        mostrarImagenDebug = 1 - mostrarImagenDebug;
		pui->desplegarImagenAR(mostrarImagenDebug);
    }

    if( key == '+' ) {
            pui->setHminApuntador(pui->getHminApuntador()+5);
            pui->setHmaxApuntador(pui->getHmaxApuntador()+5);
            printf("*** Aumento hue: [%d,%d]\n", pui->getHminApuntador(),pui->getHmaxApuntador());
	}

    if( key == '-' ) {
			pui->setHminApuntador(pui->getHminApuntador()-5);
			pui->setHmaxApuntador(pui->getHmaxApuntador()-5);
			printf("*** Disminuyo hue: [%d,%d]\n", pui->getHminApuntador(),pui->getHmaxApuntador());
	}

}

/* main loop */
static void mainLoop(void)
{
    int err_deteccion;

    if(!pui->capturarImagenAR()){
    	return;
    }else{

    	err_deteccion = pui->detectarYCalcularPosicionTodosMultiMarkers();
		if(err_deteccion==DETECCION_CORRECTA){
			idObjetoPUISeleccionado = pui->detectarSeleccion();
			if(idObjetoPUISeleccionado!=-1){
				printf("&&&&&&&&&&& COLISION CON MARCADOR %d &&&&&&&&&&&&&&&\n",idObjetoPUISeleccionado);
				correspondenciaCorrecta = false;
				correspondenciaIncorrecta = false;
				if(pui->esRelacionCorrecta(idObjetoPUIActual,idObjetoPUISeleccionado)){//FIXME no preguntar por relacion, compara ids
					correspondenciaCorrecta = true;
					actualizarPantalla();
					correspondenciaCorrecta= false;
				}else{
					correspondenciaIncorrecta=false; //FIXME tendriamos que poner correspondencia incorrecta en true
					actualizarPantalla();
				}
			}
		}else if(err_deteccion==ERROR_NINGUN_MARCADOR_DETECTADO){
			//Equivalente a marker_num<=0
		}else if(err_deteccion==ERROR_DETECTANDO_MARCADORES){
			printf("Error detectando marcadores\n");
		}else if(err_deteccion==ERROR_CALCULANDO_MATRIZ_TRANSFORMACION){
			printf("Error calculando matriz de trasnformacion\n");
		}else if(err_deteccion==ERROR_IMAGEN_NULL){
			printf("Error! imagen es NULL\n");
		}else{
			printf("err_deteccion= %d \n",err_deteccion);
		}

			if(mostrarVideo){
				if(err_deteccion!=ERROR_IMAGEN_NULL){
					pui->desplegarImagenAR(mostrarImagenDebug);
				}
    		}

			//TODO como mostramos el radar con Pui? - ver versiones anteriores donde se muestra Radar con los marcadores detectados
    	/*	if(mostrarRadar){
    			argDrawMode2D();
    			argCleanWindow(2,1);
    			dibujarLimitesPantalla(xsize, ysize);
    			for( i = 0; i < marker_num; i++ ) {
    				dibujarMiniMarcador(0,0,marker_info[i].vertex,1);
    				//printf()
    				//argDrawSquare( marker_info[i].vertex, 2, 1 );
    			}
    		}
		*/

    }
}


static void init( void ){
	printf("cargo propiedades de calibracion del apuntador desde el archivo Data/calibracionApuntadorCompleta.con\n");
	pui->initPuiConCapturaDeVideo(MODO_MULTI_MARKERS,"Data/calibracionApuntadorCompleta.con",config_name,"Data/camera_para.dat",vconf,usarAlgoritmoRapido,pui->NORMAL_SPEED_LOOP,2.0);
	pui->abrirVentanaGrafica();
	pui->capStart();



	cantElementosCargados = pui->leerConfiguracionObjetosPUI("Data/tocarAnimalesPUI/objetosPUI.config",MAX_MARCADORES);
    printf("Se cargaron %d elementos\n",cantElementosCargados);

    //Inicializo los id de las imagenes en pantalla en -1
    for(int i=0;i<cantElementosCargados;i++){
    	elementosEnPantalla[i].idImagenSDL=-1;
    }
    pui->imprimirPropiedadesObjetosPUI();


    //Inicializacion de sdl
	if(!initSDL(anchoPantalla,altoPantalla)){
		cleanUpSDL();
		pui->finish();
		delete pui;
		exit(0);
	}

	CargarFondo("Data/tocarAnimalesPUI/fondoPrototipo1.bmp",1);
	//Inicializo historial
	pui->habilitarHistorico(true,CANT_HISTORIA,-1);
	idObjetoPUIActual = 0;
	actualizarPantalla();
}

File: Biblioteca PUI/samples/python/TocarAnimalesPUI.py
#! /usr/bin/env python
# @author Sebastian Marichal
# 		   sebastian_marichal@hotmail.com
# 		   sebarocker@gmail.com
#
# Universidad de la Republica, Montevideo , Uruguay
#
import pygame, sys, os, time,gobject
sys.path.append("../../python/")
import PuiModule
from pygame import *
from pychecker.msgs import CATCH_BAD_EXCEPTION

#Configuracion del pipeline de Gstreamer
vconf = "v4l2src device=/dev/video0 ! videorate ! video/x-raw-yuv,width=320,height=240,framerate=5/1 ! ffmpegcolorspace ! capsfilter caps=video/x-raw-rgb,bpp=24 ! identity name=artoolkit ! fakesink"
#Archivo de configuracion Multi Marker de Artoolkit. Define la posicion relativa de los marcadores
config_multi_markers = "letras.dat"
#Archivo de calibracion de la camara, siempre usamos el mismo y no ha presentado problemas
cparam_name    = "../../bin/Data/camera_para.dat"
#Archivo de configuracion para la deteccion del apuntador. Este archivo fue generado mediante la utilidad CalibradorColor
calibracion_apuntador = "../../bin/Data/CalibradorColor/calibracionApuntadorCompletaPy.con"

cant_historia = 4
cantElementosCargados = -1
usarAlgoritmoRapido = 1
mostrarDebugAR = 0
idObjetoPUISeleccionado = -1
idObjetoPUIActual = -1
correspondenciaCorrecta =False

frameSize = (320,240)
mostrarVideo=1

err_deteccion = 0

thresh = 100



pui = None
dataImagen = None

running = True

#Variables de pantalla
size = (640,480)
screen = pygame.display.set_mode(size)
surface = pygame.display.get_surface()    

#
#(0,0)------(320,0)------------------------------(640,0)#
# Salida Video	|					#
#		|	Imagenes Animales		#
#-----------(320,240)					#
#		|					#
#	Radar	|					#
#		|					#
#(480,0)---------------------------------------(640,480)
def actualizarPantalla():
	global idObjetoPUIActual,surface,pui,correspondenciaCorrecta
	print '**********actualizarPantalla!!**********\n'
	if not idObjetoPUIActual==-1:
			try:
				print '%%%%%%%%%%%%%DIBUJO!%%%%%%%%%'
				img_surface =pygame.image.load(pui.getImagenObjetoPUI(idObjetoPUIActual))
				if img_surface.get_width > 320 or img_surface.get_height > 320: 
					img_surface = pygame.transform.scale(img_surface,(320,320))
				screen.blit(img_surface, (320,0))
				#idem pero con el nombre
				pygame.draw.rect(display.get_surface(), (0,255,0), (320,320,320,160),0)
				nombre_surface = pygame.font.Font(None,80).render(pui.getNombreObjetoPUI(idObjetoPUIActual),1,(255,0,0))
				screen.blit(nombre_surface,(410,380))
				pygame.display.flip()
				if correspondenciaCorrecta:
					sound = pui.getSonidoObjetoPUI(idObjetoPUIActual)
					print "****************************************Reproduccion del sonido " + sound
					pygame.mixer.music.load(pui.getSonidoObjetoPUI(idObjetoPUIActual))
					pygame.mixer.music.play()
			except Exception:
				print "Error actualizando pantalla"

def mostrarStatus():
	global screen,cantElementosCargados,pui,display
	cant_visibles = 0
	for i in range(cantElementosCargados):
		infoMarcador = pui.getInfoMultiMarcador(i)
		if infoMarcador.visible:
			cant_visibles=cant_visibles + 1
	pygame.draw.rect(display.get_surface(), (0,0,0), (40,400,80,60),0)
	
	status_surface = pygame.font.Font(None,80).render(str(cant_visibles),1,(255,0,0))
	screen.blit(status_surface,(50,400))
	pygame.display.flip()

def proximoElemento():
	global cantElementosCargados,idObjetoPUIActual
	if idObjetoPUIActual == cantElementosCargados -1:
		idObjetoPUIActual = 0
	else:
		idObjetoPUIActual = idObjetoPUIActual +1 


def desplegarDataImage(imageData):
	global running,screen,frameSize,mostrarDebugAR,mostrarVideo
	print '14'
	pg_img = pygame.image.frombuffer(imageData,frameSize, "RGB")
	print '15'	
	screen.blit(pg_img, (0,0))
	pygame.display.flip()
	print '16'
	for event in pygame.event.get():
		if event.type == pygame.QUIT:
		    running = False
		elif event.type == pygame.KEYDOWN:
			if event.key == pygame.K_ESCAPE:
				running = False
			elif event.key == K_v: #mostrar video
				mostrarVideo = 1 - mostrarVideo
			elif event.key == K_RETURN: #Proximo elemento
				proximoElemento()
				actualizarPantalla()
			elif event.key == K_t: #set threshold 
				print '*** %s (frame/sec)\n' % (pui.getFPS())
				print 'Threshold Actual = %s' % ( pui.getARThreshold() )
				user_input = input('*************** Ingrese nuevo Threshold ******************')
				print 'Threshold ingresado = %s' % (user_input)
				pui.setARThreshold(user_input)
			elif event.key == K_d: #set debug image
				mostrarDebugAR = 1 - mostrarDebugAR
				#pui.desplegarImagenAR(mostrarDebugAR)
			elif event.key == K_PLUS:
				pui.setHminApuntador(pui.getHminApuntador()+5)
				pui.setHmaxApuntador(pui.getHmaxApuntador()+5)
				print '*** Aumento hue: ['%pui.getHminApuntador()%','%pui.getHmaxApuntador()%']\n'
			elif event.key == K_MINUS: 
				pui.setHminApuntador(pui.getHminApuntador()-5)
				pui.setHmaxApuntador(pui.getHmaxApuntador()-5)
				print '*** Disminuyo hue: ['%pui.getHminApuntador()%','%pui.getHmaxApuntador()%']\n'
		else:
			print 'Evento desconocido!'

def main():
	global screen,pui,running,cantElementosCargados,cant_historia,usarAlgoritmoRapido,calibracion_apuntador,config_multi_markers,cparam_name,vconf,err_deteccion,idObjetoPUISeleccionado,idObjetoPUIActual,correspondenciaCorrecta

	#try:
	pui = PuiModule.Pui()
	#pui.setDebugMode(1)
	print '1\n'
	pui.initPuiConCapturaDeVideo(PuiModule.MODO_MULTI_MARKERS,calibracion_apuntador,config_multi_markers,cparam_name,vconf,usarAlgoritmoRapido,pui.NORMAL_SPEED_LOOP,1.0);
	
	print '2\n'	
	#Si queremos usar la funcionalidad de PUI para desplegar el video tenemos que abrir la ventana grafica
	#pui.abrirVentanaGrafica()

	pui.capStart()
	cantElementosCargados = pui.leerConfiguracionObjetosPUI("../../bin/Data/tocarAnimalesPUI/objetosPUI.config",10);
	print 'Se cargaron %s elementos\n' % cantElementosCargados

	pui.imprimirPropiedadesObjetosPUI();
	print '3\n'	
	
	idObjetoPUIActual=0
	actualizarPantalla()
	status_surface = pygame.font.Font(None,40).render('-Marcadores Visibles-',1,(255,0,0))
	screen.blit(status_surface,(20,350))
	pygame.display.flip()
	
	while running:
		dataImagen = pui.capturarPyImageAR()	#capturo imagen
		err_deteccion = pui.detectarYCalcularPosicionTodosMultiMarkers() #Detecto los marcadores y calculo las posiciones
		if err_deteccion==PuiModule.DETECCION_CORRECTA:	#No hubo error al detectar los marcadores
			idObjetoPUISeleccionado = pui.detectarSeleccion()
			if idObjetoPUISeleccionado >= 0:
				print '&&&&&&&&&&& COLISION CON MARCADOR %d &&&&&&&&&&&&&&&'  %idObjetoPUISeleccionado
				#En este caso se corrobora que el objeto PUI seleccinoado corresponda con el objeto PUI representado en 
				#pantalla. En caso de que se utilicen objetos relacionados se puede preguntar por la relacion entre ellos
				#de la siguiente manera pui.esRelacionCorrecta(idObjetoPUIActual,idObjetoPUISeleccionado)
				correspondenciaCorrecta = idObjetoPUISeleccionado==idObjetoPUIActual
				actualizarPantalla()
		#A continuacion se muestran los diferentes codigos de error que pueden ser devueltos por PUI
		#elif err_deteccion == PuiModule.ERROR_NINGUN_MARCADOR_DETECTADO:
			#no se detecto ningun marcador, eventualmente se podria alerta al usuario
			# si se reitar esta situacion varias iteraciones seguidas
			#print 'no se detectan marcadores...\n'
		#elif err_deteccion == PuiModule.ERROR_DETECTANDO_MARCADOERS:
			#error en el algoritmo de Artoolkit al detectar marcador....
		#elif err_deteccion == PuiModule.ERROR_CALCULANDO_MATRIZ_TRANSFORMACION:
			#error en el algoritmo de Artoolkit al calcular la matriz de transformacion
			#que nos da la posicion de cada marcador. Este codigo de error es muy comun
		#elif err_deteccion == PuiModule.ERROR_IMAGEN_NULL:
			#Por algun motivo la imagen es NULL, no deberia pasar nunca...
		#else:
			#error desconocido
		if not dataImagen == None:	#si se pudo obtener el frame entonces continuo
			desplegarDataImage(dataImagen.data)
			print '15'
			mostrarStatus()
			pui.capNext()
	#except Exception:
	#	print "Excepcion recibida"
					
			
if __name__ == "__main__":
	pygame.init()
	gobject.threads_init()
	main()
	pygame.quit()
	pui.finish()
	print 'Fin del programa!!'
